doc/design_intermediate_representation.md

```diff
@@ -0,0 +1,81 @@
+# _Design:_ Intermediate Representation
+
+## Overview
+
+As SQLFlow is supporting more and more machine learning toolkits, their corresponding code generation logics are better being orgnized as separate packages. An intermediate representation(IR) of the SQL jobs becomes necessary to connect these separate packages with the core `sql` package.
+
+The core `sql` package should include the following functionalities:
+1. The entry point of running extended SQL statements.
+1. The [parsing](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/sql_parser.md) of extended SQL statements.
+1. The verification of extended SQL statements, including verifying the syntax, the existence of the selected fields.
+1. The [feature derivation](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/feature_derivation.md), including name, type, shape, and preprocessing method of the select fields.
+1. The [training data and validation data split](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/training_and_validation.md).
+
+With these functionalities, the `sql` package çan translate user typed extended SQL statements to an IR as an exposed Go struct. The codegen package takes the IR and returns a generated Python program for the `sql` package to execute.
+
+## Code Structure
+
+We propose the following code structures.
+
+```
+sql/
+  ...
+  codegen/
+    tensorflow/
+      train.go
+      predict.go
+      analyze.go
+    xgboost/
+      ...
+```
+
+The `tensorflow` package will expose function `func Train(ir sql.TrainIR) string, error`, which takes the `sql`'s `TrainIR` and returns a generated Python program.
+
+## Intermediate Representation
+
+We propose the following struct as the IR for code generation.
+
+```go
+package sql
+
+// FieldMeta contains the meta information for decoding
+type FieldMeta struct {
+	DType     string // e.g. "float", "int32"
+	Delimiter string // e.g. ","
+	Shape     []int  // e.g. [1], [1 2 3]
+	IsSparse  bool   // e.g. false
+}
+
+// TrainIR is the intermediate representation for code generation of a training job
+type TrainIR struct {
+	DataSource  string                 // e.g. "hive://root:root@localhost:10000/churn"
+	ExtraConfig string                 // Extra configuration in JSON format. e.g. OSS credential
+	Select      string                 // e.g. "select * from iris.train"
+	ExtraSelect map[string]string      // e.g. {"validation": "select * from iris.val;"}
+	Estimator   string                 // e.g. "DNNClassifier"
+	Attribute   map[string]interface{} // e.g. {"train.epoch": 1000, "model.hidden_units": [10 10]}
```

Comment:
https://github.com/sql-machine-learning/sqlflow/issues/774 here's some of my thoughts about refactoring attribute resolution.

Reply:
@typhoonzero Thanks for the link. I think `Attribute map[string]interface{}` here is fine.

I believe `AttrMeta` belong to the code generation package since different machine learning toolkits accept different attributes.

Reply:
> I believe AttrMeta belong to the code generation package since different machine learning toolkits accept different attributes

It is. Each submitter's code generation package define it's `AttrMeta` and call a function to get an `Attribute map[string]interface{}`.